File: tasks/2977.py
import math

INF = math.inf


class Solution:
    def minimumCost(self, source: str, target: str, original: list[str], changed: list[str], cost: list[int]) -> int:
        substrings = list(set(original) | set(changed))
        substr_id = {substring: i for i, substring in enumerate(substrings)}  # нумеруем все подстроки
        substr_num = len(substrings)

        # строим граф замен подстрок: узлы - подстроки, ребра с минимальной стоимостью - доступные замены
        costs = [[INF] * substr_num for _ in range(substr_num)]
        for i in range(substr_num):
            costs[i][i] = 0

        # сначала просто записываем замены
        for i in range(len(original)):
            old, new, c = original[i], changed[i], cost[i]
            j, k = substr_id[old], substr_id[new]
            if c < costs[j][k]:
                costs[j][k] = c

        # теперь ищем минимальные пути из i в j через все доступные k - алгоритм Флойда-Уоршелла
        for k in range(substr_num):
            cost_k = costs[k]
            for i in range(substr_num):
                cost_i = costs[i]
                if cost_i[k] == INF:
                    continue
                for j in range(substr_num):
                    if cost_i[k] + cost_k[j] < cost_i[j]:
                        cost_i[j] = cost_i[k] + cost_k[j]

        lengths = sorted(list(set(len(s) for s in original)))  # записываем длины подстрок, которые можно заменить

        num = len(source)
        dynamic_costs = [INF] * (num + 1)  # динамический массив для нахождения оптимальных цен замен подстрок
        dynamic_costs[0] = 0

        for i in range(num):
            if dynamic_costs[i] == INF:  # не можем дойти до этого символа
                continue
            if source[i] == target[i]:  # данный символ совпадает с целью, шагаем вперед бесплатно
                if dynamic_costs[i] < dynamic_costs[i + 1]:
                    dynamic_costs[i + 1] = dynamic_costs[i]

            for length in lengths:  # перебираем всевозможные варианты длин подстрок
                if i + length > num:
                    break
                old = source[i: i + length]
                new = target[i: i + length]

                j = substr_id.get(old)
                k = substr_id.get(new)

                if j is not None and k is not None:
                    cost_jk = costs[j][k]
                    if cost_jk != INF:  # находим возможную замену подстроки и сравниваем цену
                        if dynamic_costs[i] + cost_jk < dynamic_costs[i + length]:
                            dynamic_costs[i + length] = dynamic_costs[i] + cost_jk

        return dynamic_costs[num] if dynamic_costs[num] != INF else -1

File: src/2977.kt
import kotlin.math.min

class Solution {
    fun minimumCost(source: String, target: String, original: Array<String>, changed: Array<String>, cost: IntArray): Long {
        val n = source.length
        val INF = 1e15.toLong()
        val rulesByLen = mutableMapOf<Int, MutableList<Triple<String, String, Int>>>()
        for (i in original.indices) {
            rulesByLen.getOrPut(original[i].length) { mutableListOf() }.add(Triple(original[i], changed[i], cost[i]))
        }
        val costMap = mutableMapOf<Int, Map<String, Map<String, Long>>>()
        for ((len, triples) in rulesByLen) {
            val nodes = triples.flatMap { listOf(it.first, it.second) }.toSet().toList()
            val idMap = nodes.withIndex().associate { it.value to it.index }
            val m = nodes.size
            val dist = Array(m) { LongArray(m) { INF } }
            for (i in 0 until m) dist[i][i] = 0L
            for ((u, v, c) in triples) {
                val i = idMap[u]!!
                val j = idMap[v]!!
                dist[i][j] = min(dist[i][j], c.toLong())
            }
            for (k in 0 until m) {
                for (i in 0 until m) {
                    if (dist[i][k] == INF) continue
                    for (j in 0 until m) {
                        if (dist[k][j] == INF) continue
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
                    }
                }
            }
            val resMap = mutableMapOf<String, MutableMap<String, Long>>()
            for (i in 0 until m) {
                for (j in 0 until m) {
                    if (dist[i][j] != INF) {
                        resMap.getOrPut(nodes[i]) { mutableMapOf() }[nodes[j]] = dist[i][j]
                    }
                }
            }
            costMap[len] = resMap
        }
        val dp = LongArray(n + 1) { INF }
        dp[0] = 0L
        val lengths = costMap.keys.sorted()
        for (i in 1..n) {
            if (source[i - 1] == target[i - 1]) {
                dp[i] = dp[i - 1]
            }
            for (len in lengths) {
                if (i >= len && dp[i - len] != INF) {
                    val s = source.substring(i - len, i)
                    val t = target.substring(i - len, i)
                    val c = costMap[len]?.get(s)?.get(t) ?: INF
                    if (c != INF) {
                        dp[i] = min(dp[i], dp[i - len] + c)
                    }
                }
            }
        }
        return if (dp[n] >= INF) -1L else dp[n]
    }
}
